gateway/ln-gateway/src/lightning/lnd.rs

```diff
@@ -464,73 +473,48 @@ impl ILnRpcClient for GatewayLndClient {
                     }
-                })?;
+                })?
+                .into_inner();
 
-            debug!(
-                "LND payment request sent for invoice {invoice:?}, waiting for payment status..."
-            );
-            let mut messages = payments.into_inner();
-            loop {
-                match messages
-                    .message()
-                    .await
-                    .map_err(|error| LightningRpcError::FailedPayment {
-                        failure_reason: format!("Failed to get payment status {error:?}"),
-                    }) {
-                    Ok(Some(payment)) if payment.status() == PaymentStatus::Succeeded => {
-                        info!("LND payment succeeded for invoice {invoice:?}");
-                        break hex::FromHex::from_hex(payment.payment_preimage.as_str()).map_err(
-                            |error| LightningRpcError::FailedPayment {
-                                failure_reason: format!("Failed to convert preimage {error:?}"),
-                            },
-                        )?;
-                    }
-                    Ok(Some(payment)) if payment.status() == PaymentStatus::InFlight => {
-                        debug!("LND payment is inflight");
-                        continue;
-                    }
-                    Ok(Some(payment)) => {
-                        info!("LND payment failed for invoice {invoice:?} with {payment:?}");
-                        let failure_reason = payment.failure_reason();
-                        return Err(LightningRpcError::FailedPayment {
-                            failure_reason: format!("{failure_reason:?}"),
-                        });
-                    }
-                    Ok(None) => {
-                        info!("LND payment failed for invoice {invoice:?} with no payment status");
-                        return Err(LightningRpcError::FailedPayment {
-                            failure_reason: format!(
-                                "Failed to get payment status for payment hash {:?}",
-                                invoice.payment_hash
-                            ),
-                        });
-                    }
-                    Err(e) => {
-                        info!("LND payment failed for invoice {invoice:?} with {e:?}");
-                        return Err(e);
-                    }
+            let sleep_fut = sleep(Duration::from_secs(LND_PAYMENT_TIMEOUT_SECONDS));
+            let payment_result_fut = attempt_private_payments(routes, invoice, &mut client);
+            pin_mut!(sleep_fut, payment_result_fut);
+
+            match futures::future::select(sleep_fut, payment_result_fut).await {
+                Either::Left(_) => {
+                    return Err(LightningRpcError::FailedPayment {
+                        failure_reason: "Timed out waiting for payment to complete".to_string(),
```

Comment:
More of a general question not specific to this PR: how should we handle stuck payments that time out here?

If I recall correctly, a timeout here will cause the gateway to cancel the payment and the user to get a refund. But that opens the door for the user getting refunded and the payment still going through, causing the gateway to lose money. Perhaps when this happens, the gateway should not cancel the payment and force the user to get refunded by waiting for the timeout :/

Core lightning has the same problem: https://github.com/fedimint/fedimint/pull/5449